lib/FusionInventory/Agent/Tools/Hardware.pm

```diff
@@ -540,18 +559,56 @@ sub _getMacAddress {
     # fallback on ports addresses (IF-MIB::ifPhysAddress) if unique
     my $addresses_oid = ".1.3.6.1.2.1.2.2.1.6";
     my $addresses = $snmp->walk($addresses_oid);
+
+    # filter keeping mac with a defined and associated ip otherwise takes all macs
+    my $ips = $snmp->walk('.1.3.6.1.2.1.4.20.1.2');
+    my @macs = ();
+    @macs = grep { defined } map { $addresses->{$_} } values %{$ips}
+        if (keys(%{$ips}));
+    @macs = grep { defined } values %{$addresses}
+        unless @macs;
+
     my @addresses =
         uniq
+        grep { /^$mac_address_pattern$/ }
         grep { $_ ne '00:00:00:00:00:00' }
         grep { $_ }
         map  { _getCanonicalMacAddress($_) }
-        values %{$addresses};
+        @macs;
+
+    if (@addresses) {
+        return $addresses[0] if @addresses == 1;
 
-    return $addresses[0] if @addresses && @addresses == 1;
+        # Compute mac addresses as number and sort them
+        my %macs = map { $_ => _numericMac($_) } @addresses;
```

Comment:
Yes this is necessary as this is not only to sort addresses but also to test on mac address is exactly following another one as tested line 590.

The idea is to associate the first mac address of a series as the device mac address as manufacturers always affects a series when the device needs some mac addresses.